modules/renter/repairscanner.go

```diff
@@ -0,0 +1,333 @@
+	// structures.
+	index  uint64
+	length uint64
+	offset int64
+
+	// The logical data is the data that is presented to the user when the user
+	// requests the chunk. The physical data is all of the pieces that get
+	// stored across the network.
+	logicalChunkData  []byte
+	physicalChunkData [][]byte
+
+	// Fields for tracking the current progress of the chunk and all the pieces.
+	memoryNeeded     uint64 // memory needed in bytes
+	piecesNeeded     int // number of pieces to achieve a 100% complete upload
+	piecesCompleted  int // number of pieces that have been fully uploaded.
+	piecesRegistered int // number of pieces that are being uploaded, but aren't finished yet.
+	pieceUsage       []bool // one per piece. 'false' = piece not uploaded. 'true' = piece uploaded.
+	progress         float64 // percent complete at the moment uploading started
+	unusedHosts      map[string]struct{} // hosts that aren't yet storing any pieces
+
+	// Utilities.
+	mu sync.Mutex
+}
+
+func (ch chunkHeap) Len() int            { return len(ch) }
+func (ch chunkHeap) Less(i, j int) bool  { return ch[i].progress < ch[j].progress }
+func (ch chunkHeap) Swap(i, j int)       { ch[i], ch[j] = ch[j], ch[i] }
+func (ch *chunkHeap) Push(x interface{}) { *ch = append(*ch, x.(*unfinishedChunk)) }
+func (ch *chunkHeap) Pop() interface{} {
+	old := *ch
+	n := len(old)
+	x := old[n-1]
+	*ch = old[0 : n-1]
+	return x
+}
+
+// buildUnfinishedChunks will pull all of the unfinished chunks out of a file.
+//
+// TODO / NOTE: This code can be substantially simplified once the files store
+// the HostPubKey instead of the FileContractID, and can be simplified even
+// further once the layout is per-chunk instead of per-filecontract.
+func (r *Renter) buildUnfinishedChunks(f *file, hosts map[string]struct{}, fcidToHPK map[types.FileContractID]types.SiaPublicKey) []*unfinishedChunk {
+	// Files are not threadsafe.
+	f.mu.Lock()
+	defer f.mu.Unlock()
+
+	// If the file is not being tracked, don't repair it.
+	trackedFile, exists := r.tracking[f.name]
+	if !exists {
+		return nil
+	}
+
+	// Assemble the set of chunks.
+	//
+	// TODO / NOTE: Future files may have a different method for determining the
+	// number of chunks. Changes will be made due to things like sparse files,
+	// and the fact that chunks are going to be different sizes.
+	chunkCount := f.numChunks()
+	newUnfinishedChunks := make([]*unfinishedChunk, chunkCount)
+	// Add a separate unusedHosts map for each chunk, as every chunk will have a
+	// different set of unused hosts.
+	for i := uint64(0); i < chunkCount; i++ {
+		newUnfinishedChunks[i] = new(unfinishedChunk)
+		newUnfinishedChunks[i].index = i
+		newUnfinishedChunks[i].localPath = trackedFile.RepairPath
+
+		// Mark the number of pieces needed for this chunk.
+		newUnfinishedChunks[i].piecesNeeded = f.erasureCode.NumPieces()
+		newUnfinishedChunks[i].memoryNeeded = f.pieceSize * uint64(f.erasureCode.NumPieces())
```

Comment:
my preferred style here would be to start with
```go
c := &unfinishedChunk{
    index:        i,
    localPath:    trackedFile.RepairPath,
    piecesNeeded: f.erasureCode.NumPieces(),
    memoryNeeded: f.pieceSize * uint64(f.erasureCode.NumPieces()),
}
```
and assign `c` to `newUnfinishedChunks[i]` at the end of the loop. Less verbose that way